runtime/compiler/optimizer/EscapeAnalysis.cpp

```diff
@@ -3742,32 +3742,60 @@ static bool isFinalizableInlineTest(TR::Compilation *comp, TR::Node *candidate,
+   TR::Node *r1 = (root->getNumChildren() > 0) ? root->getFirstChild() : NULL;
+   TR::Node *r2 = (root->getNumChildren() > 1) ? root->getSecondChild() : NULL;
+   TR::Node *r11 = (r1 && r1->getNumChildren() > 0) ? r1->getFirstChild() : NULL;
+   TR::Node *r12 = (r1 && r1->getNumChildren() > 1) ? r1->getSecondChild() : NULL;
+
+   bool castDownToInt = is64Bit && r11 && (r11->getOpCodeValue() == TR::l2i);
+   bool usesLongOps = is64Bit && !castDownToInt;
+
+   TR::ILOpCodes ifOp = usesLongOps ? TR::iflcmpne : TR::ificmpne;
+   TR::ILOpCodes andOp = usesLongOps ? TR::land : TR::iand;
    TR::ILOpCodes loadOp = is64Bit ? TR::lloadi : TR::iloadi;
-   TR::ILOpCodes constOp = is64Bit ? TR::lconst : TR::iconst;
+   TR::ILOpCodes constOp = usesLongOps ? TR::lconst : TR::iconst;
+
+   TR::Node *loadNode = castDownToInt ? r11->getFirstChild() : r11;
 
    /*
       Looking for a pattern of:
+
       if[i/l]cmpne                                      <root>
-         [i/l]and
-            [i/l]loadi <classAndDepthFlags>
+         [i/l]and                                       <r1>
+            [i/l]loadi <classAndDepthFlags>             <r11/loadNode>
                aloadi <vft-symbol>                      <vftLoad>
                   ...                                   <ref>
-            [i/l]const <FlagValueForFinalizerCheck>
-         [i/l]const 0
+            [i/l]const <FlagValueForFinalizerCheck>     <r12>
+         [i/l]const 0                                   <r2>
+
+       or
+
+      ificmpne                                          <root>
+         iand                                           <r1>
+            l2i                                         <r11>
+               lloadi <classAndDepthFlags>              <loadNode>
+                  aloadi <vft-symbol>                   <vftLoad>
+                     ...                                <ref>
+            iconst <FlagValueForFinalizerCheck>         <r12>
+         iconst 0                                       <r2>
    */
 
    return root->getOpCodeValue() == ifOp &&
-          root->getFirstChild()->getOpCodeValue() == andOp &&
-          root->getSecondChild()->getOpCodeValue() == constOp &&
-          (is64Bit ? root->getSecondChild()->getLongInt() == 0 : root->getSecondChild()->getInt() == 0) &&
-          root->getFirstChild()->getFirstChild()->getOpCodeValue() == loadOp &&
-          root->getFirstChild()->getSecondChild()->getOpCodeValue() == constOp &&
-          (is64Bit ?
-           root->getFirstChild()->getSecondChild()->getLongInt() == fej9->getFlagValueForFinalizerCheck() :
-           root->getFirstChild()->getSecondChild()->getInt() == fej9->getFlagValueForFinalizerCheck()) &&
-          root->getFirstChild()->getFirstChild()->getFirstChild() == vftLoad /*&& (implied by the above assume)
+          r1->getOpCodeValue() == andOp &&
+          r2->getOpCodeValue() == constOp &&
+          (usesLongOps ? r2->getLongInt() == 0 : r2->getInt() == 0) &&
```

Comment:
would this be more clear as:

`(usesLongOps ? r2->getLongInt() : r2->getInt()) == 0` - that would imply some kind of type conversion but that wouldn't affect the value 0 which is what we would be looking for...